File: backend/internal/domain/furniture_definition.go
package domain

import "encoding/json"

// FurnitureDefinition (Layer 2) is design intent only — no final geometry or parts.

type FurnitureDefinition struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Description string          `json:"description,omitempty"`
	Root        VolumeNode      `json:"root"`
	Metadata    json.RawMessage `json:"metadata,omitempty"`
}

type VolumeNode struct {
	ID             string              `json:"id"`
	Label          string              `json:"label,omitempty"`
	Constraints    VolumeConstraints   `json:"constraints"`
	Split          *VolumeSplit        `json:"split,omitempty"`
	Children       []VolumeNode        `json:"children"`
	Features       []Feature           `json:"features"`
	Fronts         []Front             `json:"fronts"`
	Adaptation     *AdaptationRules    `json:"adaptation,omitempty"`
	Manufacturing  *ManufacturingHints `json:"manufacturing,omitempty"`
}

type VolumeConstraints struct {
	Width  *DimensionConstraint `json:"width,omitempty"`
	Height *DimensionConstraint `json:"height,omitempty"`
	Depth  *DimensionConstraint `json:"depth,omitempty"`
}

type DimensionMode string

const (
	DimensionFixed DimensionMode = "fixed"
	DimensionRatio DimensionMode = "ratio"
	DimensionFill  DimensionMode = "fill"
	DimensionMin   DimensionMode = "min"
	DimensionMax   DimensionMode = "max"
)

type DimensionConstraint struct {
	Mode  DimensionMode `json:"mode"`
	Value float64       `json:"value,omitempty"`
}

type SplitAxis string

const (
	SplitAxisX SplitAxis = "x"
	SplitAxisY SplitAxis = "y"
	SplitAxisZ SplitAxis = "z"
)

type VolumeSplit struct {
	Axis   SplitAxis `json:"axis"`
	Ratios []float64 `json:"ratios,omitempty"`
	Fixed  []float64 `json:"fixed,omitempty"`
}

// Feature is extensible — type is a string, not a closed enum.
type Feature struct {
	ID     string          `json:"id"`
	Type   string          `json:"type"`
	Params json.RawMessage `json:"params"`
}

type Front struct {
	ID     string          `json:"id"`
	Type   string          `json:"type"`
	Params json.RawMessage `json:"params"`
}

type AdaptationRules struct {
	FollowFloor         bool            `json:"followFloor,omitempty"`
	FollowCeiling       bool            `json:"followCeiling,omitempty"`
	FollowWall          bool            `json:"followWall,omitempty"`
	CompensateSkirting  bool            `json:"compensateSkirting,omitempty"`
	Params              json.RawMessage `json:"params,omitempty"`
}

type ManufacturingHints struct {
	MaterialID   string          `json:"materialId,omitempty"`
	EdgeBanding  string          `json:"edgeBanding,omitempty"`
	BackPanel    bool            `json:"backPanel,omitempty"`
	Params       json.RawMessage `json:"params,omitempty"`
}

File: frontend/src/domain/furniture-definition.ts
/**
 * Layer 2 — Furniture Definition
 * Design intent only. No final geometry, cuts, or parts.
 */

export interface FurnitureDefinition {
  id: string;
  name: string;
  description?: string;
  root: VolumeNode;
  metadata?: Record<string, unknown>;
}

export interface VolumeNode {
  id: string;
  label?: string;
  constraints: VolumeConstraints;
  split?: VolumeSplit;
  children: VolumeNode[];
  features: Feature[];
  fronts: Front[];
  adaptation?: AdaptationRules;
  manufacturing?: ManufacturingHints;
}

export interface VolumeConstraints {
  width?: DimensionConstraint;
  height?: DimensionConstraint;
  depth?: DimensionConstraint;
}

export type DimensionConstraint =
  | { mode: 'fixed'; value: number }
  | { mode: 'ratio'; value: number }
  | { mode: 'fill' }
  | { mode: 'min'; value: number }
  | { mode: 'max'; value: number };

export type SplitAxis = 'x' | 'y' | 'z';

export interface VolumeSplit {
  axis: SplitAxis;
  /** Relative proportions for child volumes (sum should equal 1). */
  ratios?: number[];
  /** Fixed sizes in mm for children that are not ratio-based. */
  fixed?: number[];
}

/**
 * Extensible feature — never use a closed enum of all furniture types.
 * Examples: shelf_set, drawer_stack, hanger_rod, divider, lighting, appliance_space
 */
export interface Feature {
  id: string;
  type: string;
  params: Record<string, unknown>;
}

/**
 * Visible front elements.
 * Examples: door, sliding_door, glass, mirror, drawer_front
 */
export interface Front {
  id: string;
  type: string;
  params: Record<string, unknown>;
}

export interface AdaptationRules {
  followFloor?: boolean;
  followCeiling?: boolean;
  followWall?: boolean;
  compensateSkirting?: boolean;
  params?: Record<string, unknown>;
}

export interface ManufacturingHints {
  materialId?: string;
  edgeBanding?: string;
  backPanel?: boolean;
  params?: Record<string, unknown>;
}
